File: trunk/util/sim-event-processor/src/edu/colorado/phet/simeventprocessor/EventLog.java
// Copyright 2002-2011, University of Colorado
package edu.colorado.phet.simeventprocessor;

import java.io.File;
import java.util.ArrayList;
import java.util.Date;
import java.util.Iterator;
import java.util.List;
import java.util.StringTokenizer;

import edu.colorado.phet.common.phetcommon.util.ObservableList;
import edu.colorado.phet.common.phetcommon.util.Option;
import edu.colorado.phet.common.phetcommon.util.function.Function1;

/**
 * State for a single parse run.
 *
 * @author Sam Reid
 */
public class EventLog implements Iterable<Entry> {
    private String machineID;
    private String sessionID;
    private long serverTime;
    public List<Entry> lines = new ArrayList<Entry>();
    public final File sourceFile;

    public EventLog( File sourceFile ) {
        this.sourceFile = sourceFile;
    }

    public EventLog( String machineID, String sessionID, long serverTime, List<Entry> lines, File sourceFile ) {
        this.machineID = machineID;
        this.sessionID = sessionID;
        this.serverTime = serverTime;
        this.lines = lines;
        this.sourceFile = sourceFile;
    }

    //Parse a single line
    public void parseLine( String line ) {
        if ( line.startsWith( "machineID" ) ) {
            machineID = readValue( line );
        }
        else if ( line.startsWith( "sessionID" ) ) {
            sessionID = readValue( line );
        }
        else if ( line.startsWith( "serverTime" ) ) {
            serverTime = Long.parseLong( readValue( line ) );
        }
        else {
            Entry entry = Entry.parse( line );
            lines.add( entry );
        }
    }

    //Read the third token, the value in a "name = value" list
    private String readValue( String line ) {
        StringTokenizer stringTokenizer = new StringTokenizer( line, " " );
        stringTokenizer.nextToken();//machineID
        stringTokenizer.nextToken();//=
        return stringTokenizer.nextToken();
    }

    public Iterator<Entry> iterator() {
        return lines.iterator();
    }

    public EventLog getWithoutSystemEvents() {
        return removeItems( new Function1<Entry, Boolean>() {
            public Boolean apply( Entry entry ) {
                return entry.actor.toLowerCase().startsWith( "system" );
            }
        } );
    }

    public EventLog removeItems( Function1<Entry, Boolean> filter ) {
        return new EventLog( machineID, sessionID, serverTime, new ObservableList<Entry>( lines ).removeItems( filter ), sourceFile );
    }

    public int getLastTime() {
        return (int) lines.get( lines.size() - 1 ).timeMilliSec;
    }

    public int getNumberOfEvents( final long time ) {
        return getNumberOfEvents( time, new Function1.Constant<Entry, Boolean>( true ) );
    }

    public int getNumberOfEvents( final long time, Function1<Entry, Boolean> matches ) {
        EventLog log = removeItems( new Function1<Entry, Boolean>() {
            public Boolean apply( Entry entry ) {
                return entry.timeMilliSec > time;
            }
        } );
        EventLog user = log.getWithoutSystemEvents();
        EventLog keep = user.keepItems( matches );
        return keep.size();
    }

    public EventLog keepItems( Function1<Entry, Boolean> matches ) {
        return new EventLog( machineID, sessionID, serverTime, new ObservableList<Entry>( lines ).keepItems( matches ), sourceFile );
    }

    public int size() {
        return lines.size();
    }

    //How many events in the list happened in the log
    public int getNumberOfEvents( final long time, EntryList eventsOfInterest ) {
        EventLog log = removeItems( new Function1<Entry, Boolean>() {
            public Boolean apply( Entry entry ) {
                return entry.timeMilliSec > time;
            }
        } );
        EventLog user = log.getWithoutSystemEvents();
        int count = 0;
        for ( Entry eventOfInterest : eventsOfInterest ) {
            if ( user.containsMatch( eventOfInterest ) ) {
                count++;
            }
        }
        return count;
    }

    private boolean containsMatch( Entry event ) {
        for ( Entry line : lines ) {
            if ( line.matches( event.actor, event.event, event.parameters ) ) { return true; }
        }
        return false;
    }

    //Which of the specified events of interest are in our list?
    public EntryList find( EntryList eventsOfInterest ) {
        EntryList list = new EntryList();
        for ( Entry entry : eventsOfInterest ) {
            if ( containsMatch( entry ) ) {
                list.add( entry );
            }
        }
        return list;
    }

    public long getServerStartTime() {
        return serverTime;
    }

    public String brief() {
        final Entry startMessage = getStartMessage();
        return startMessage.get( "name" ) + " " + startMessage.get( "version" ).get() + " startTime = " + new Date( serverTime ) + ", epoch = " + serverTime + ", study = " + startMessage.get( "study" ) + ", userID = " + startMessage.get( "id" ) + ", events = " + size() + ", timeUsed = " + minutesUsed() + " minutes, machineID = " + machineID + ", sessionID = " + sessionID;
    }

    public int minutesUsed() {
        return getLastTime() / 1000 / 60;
    }

    private Entry getStartMessage() {
        return getFirstEntry( "system", "started" );
    }

    private Entry getFirstEntry( String system, String started ) {
        for ( Entry line : lines ) {
            if ( line.matches( system, started ) ) {
                return line;
            }
        }
        return null;
    }

    public Option<String> getID() {
        return getStartMessage().get( "id" );
    }

    public String getSimName() {
        return getStartMessage().get( "name" ).get();
    }

    public String getMachineID() {
        return machineID;
    }

    public String getStudy() {
        return getStartMessage().get( "study" ).get();
    }

    @Override public String toString() {
        return "EventLog{" +
               "machineID='" + machineID + '\'' +
               ", sessionID='" + sessionID + '\'' +
               ", serverTime=" + serverTime +
               ", lines=" + lines +
               ", sourceFile=" + sourceFile +
               '}';
    }
}

File: trunk/util/sim-event-processor/src/edu/colorado/phet/simeventprocessor/scala/EventLog.scala
// Copyright 2002-2011, University of Colorado
package edu.colorado.phet.simeventprocessor.scala

import java.util.Date
import java.text.SimpleDateFormat
import collection.mutable.HashMap
import scala.collection.JavaConversions._
import edu.colorado.phet.simeventprocessor.{JavaEntry, JavaEventLog}
import java.lang.Boolean
import edu.colorado.phet.common.phetcommon.simsharing.Parameter

/**
 * Adds scala-convenient interface for REPL.
 * @author Sam Reid
 */
class EventLog(log: JavaEventLog) {
  val machine = log.getMachineID
  val simName = log.getSimName
  val epoch = log.getServerStartTime
  val simVersion = log.getSimVersion
  val study = log.getStudy
  val size = log.size
  val session = log.getSessionID
  val user = log.getID
  val startDate = new Date(epoch)
  val day = new SimpleDateFormat("MM-dd-yyyy").format(startDate)
  val lastTime = log.getLastTime
  val osName = log.getOSName
  val osVersion = log.getOSVersion

  def selectFirstTab(tab: String) = {
    val a = new JavaEventLog(log.file)

    var recording = true
    for ( entry <- entries ) {
      if ( recording ) {
        a.lines.add(entry)
      }
      if ( entry.matches("tab", "pressed", List(new Parameter("text", tab)).toArray) ) {
        recording = true
      }
      else if ( entry.matches("tab", "pressed") ) {
        recording = false
      }
    }
    new EventLog(a)
  }

  //Only looks for subsequent tabs, have to do something else for first tab
  def selectLaterTab(tab: String) = {
    val a = new JavaEventLog(log.file)

    //Add sim started message to get sim name etc
    a.lines.add(entries(0))

    var recording = false
    for ( entry <- entries ) {
      if ( recording ) {
        a.lines.add(entry)
      }
      if ( entry.matches("tab", "pressed", List(new Parameter("text", tab)).toArray) ) {
        recording = true
      }
      else if ( entry.matches("tab", "pressed") ) {
        recording = false
      }
    }
    new EventLog(a)
  }

  val entries = {
    ( for ( elm <- log ) yield {
      elm
    } ).toList
  }
  lazy val userNumber = {
    try {
      user.toInt
    }
    catch {
      case nfe: NumberFormatException => -1;
    }
  }

  //Millis
  lazy val firstUserEvent = entries.find(log => log.actor != "system" && log.actor != "window")

  def countEvents(until: Long): Int = log.getNumberOfEvents(until)

  def countMatches(matcher: Seq[Match], until: Long): Int = {
    val earlyEnoughEvents = log.filter(_.timeMilliSec < until)
    matcher.filter(matchItem => earlyEnoughEvents.find(matchItem.matches(_)).isDefined).size
  }

  override def toString = simName + " " + new Date(epoch) + " (" + epoch + "), study = " + study + ", user = " + user + ", events = " + size + ", machine = " + machine + ", session = " + session

  lazy val histogramByObject = {
    val map = new HashMap[String, Int]
    for ( elm: JavaEntry <- log ) {
      val currentValue = map.getOrElse(elm.actor, 0)
      map.put(elm.actor, currentValue + 1)
    }
    map
  }

  def matches(entry: Match): Boolean = log.filter(entry.matches(_)).size > 0

  def findMatches(matcher: Seq[Match]): List[Match] = ( for ( m <- matcher if matches(m) ) yield {
    m
  } ).toList

  def findEvents(matcher: Match): List[JavaEntry] = log.filter(matcher.matches(_)).toList

  def findEvents(actor: String): List[JavaEntry] = log.filter(ActorRule(actor).matches(_)).toList

  def foreach[U](f: ( JavaEntry ) => U) {
    for ( elm <- log ) {
      f(elm)
    }
  }

  lazy val eventCountData = phet.timeSeries(this, countEvents(_))

  //  val importantEvents = simLogs.map(log => timeSeries(log, log.countMatches(simEventMap(sim), _)))
  def countEvents(matcher: Seq[Match]) = phet.timeSeries(this, countMatches(matcher, _))

  private def containsParameters(e: JavaEntry, pairs: Seq[Pair[String, String]]): Boolean = {
    for ( p <- pairs ) {
      if ( !e.hasParameter(p._1, p._2) ) {
        return false
      }
    }
    true
  }

  def contains(actor: String, event: String, pairs: Pair[String, String]*) = log.find((e: JavaEntry) => e.actor == actor && e.event == event && containsParameters(e, pairs)).isDefined

}